File: mobile/test/paths_test.dart
/// Guards the untrusted-filename handling in lib/receive/paths.dart.
///
/// The filename arrives from another machine, so traversal and collision are
/// the two ways this hurts: escaping the save directory, or silently eating a
/// file you already caught. Mirrors the intent of config.safe_filename /
/// config.unique_path on the desktop.
/// Run: flutter test
library;

import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:yoink_mobile/receive/paths.dart';

void main() {
  group('safeFilename', () {
    test('keeps ordinary names', () {
      expect(safeFilename('report.pdf'), 'report.pdf');
      expect(safeFilename('holiday photo.png'), 'holiday photo.png');
    });

    test('strips any path component', () {
      expect(safeFilename('../../etc/passwd'), 'passwd');
      expect(safeFilename(r'C:\windows\system32\evil.dll'), 'evil.dll');
      expect(safeFilename('/data/data/other.app/db'), 'db');
      expect(safeFilename(r'a/b\c/d.txt'), 'd.txt');
    });

    test('falls back when nothing usable survives', () {
      expect(safeFilename(null), 'yoink-payload');
      expect(safeFilename(''), 'yoink-payload');
      expect(safeFilename('..'), 'yoink-payload');
      expect(safeFilename('.'), 'yoink-payload');
      expect(safeFilename('...'), 'yoink-payload');
      expect(safeFilename('/'), 'yoink-payload');
      expect(safeFilename('x', fallback: 'yoink-image'), 'x');
      expect(safeFilename('..', fallback: 'yoink-image'), 'yoink-image');
    });

    test('scrubs characters a filesystem rejects', () {
      expect(safeFilename('a<b>c:d"e|f?g*h.txt'), 'a_b_c_d_e_f_g_h.txt');
      expect(safeFilename('bell\x07.txt'), 'bell_.txt'); // control chars too
    });

    test('trims trailing dots and spaces', () {
      expect(safeFilename('name.   '), 'name');
      expect(safeFilename('name...'), 'name');
    });

    test('caps the length', () {
      expect(safeFilename('${'a' * 500}.png').length, 120);
    });

    test('the result can never escape its directory', () {
      const nasty = [
        '../../../etc/passwd',
        r'..\..\windows\evil.exe',
        '....//....//x',
        '/absolute/path',
        '..',
      ];
      for (final n in nasty) {
        final safe = safeFilename(n);
        expect(safe, isNot(contains('/')), reason: n);
        expect(safe, isNot(contains(r'\')), reason: n);
        expect(safe, isNot('..'), reason: n);
      }
    });
  });

  group('uniquePath', () {
    late Directory dir;

    setUp(() => dir = Directory.systemTemp.createTempSync('yoink_test'));
    tearDown(() => dir.deleteSync(recursive: true));

    test('uses the plain name when it is free', () {
      expect(uniquePath(dir, 'a.png').path, '${dir.path}/a.png');
    });

    test('never overwrites an existing catch', () {
      File('${dir.path}/photo.png').writeAsStringSync('first');
      final second = uniquePath(dir, 'photo.png');
      expect(second.path, '${dir.path}/photo (1).png');

      second.writeAsStringSync('second');
      expect(uniquePath(dir, 'photo.png').path, '${dir.path}/photo (2).png');
      // The original survived, which is the whole point.
      expect(File('${dir.path}/photo.png').readAsStringSync(), 'first');
    });

    test('handles names with no extension', () {
      File('${dir.path}/README').writeAsStringSync('x');
      expect(uniquePath(dir, 'README').path, '${dir.path}/README (1)');
    });

    test('treats a leading dot as part of the name, not an extension', () {
      File('${dir.path}/.gitignore').writeAsStringSync('x');
      expect(uniquePath(dir, '.gitignore').path, '${dir.path}/.gitignore (1)');
    });
  });
}

File: android/app/src/test/kotlin/com/yoink/receive/PathsTest.kt
package com.yoink.receive

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

/**
 * Guards the untrusted-filename handling in Paths.kt (port of the Dart
 * paths_test.dart). Traversal and collision are the two ways this hurts:
 * escaping the save directory, or silently eating a file you already caught.
 * Run: gradlew :app:testDebugUnitTest
 */
class PathsTest {

    @get:Rule
    val tmp = TemporaryFolder()

    @Test
    fun keepsOrdinaryNames() {
        assertEquals("report.pdf", Paths.safeFilename("report.pdf"))
        assertEquals("holiday photo.png", Paths.safeFilename("holiday photo.png"))
    }

    @Test
    fun stripsAnyPathComponent() {
        assertEquals("passwd", Paths.safeFilename("../../etc/passwd"))
        assertEquals("evil.dll", Paths.safeFilename("""C:\windows\system32\evil.dll"""))
        assertEquals("db", Paths.safeFilename("/data/data/other.app/db"))
        assertEquals("d.txt", Paths.safeFilename("""a/b\c/d.txt"""))
    }

    @Test
    fun fallsBackWhenNothingUsableSurvives() {
        assertEquals("yoink-payload", Paths.safeFilename(null))
        assertEquals("yoink-payload", Paths.safeFilename(""))
        assertEquals("yoink-payload", Paths.safeFilename(".."))
        assertEquals("yoink-payload", Paths.safeFilename("."))
        assertEquals("yoink-payload", Paths.safeFilename("..."))
        assertEquals("yoink-payload", Paths.safeFilename("/"))
        assertEquals("x", Paths.safeFilename("x", "yoink-image"))
        assertEquals("yoink-image", Paths.safeFilename("..", "yoink-image"))
    }

    @Test
    fun scrubsCharactersAFilesystemRejects() {
        assertEquals("a_b_c_d_e_f_g_h.txt", Paths.safeFilename("""a<b>c:d"e|f?g*h.txt"""))
        assertEquals("bell_.txt", Paths.safeFilename("bell\u0007.txt")) // control chars too
    }

    @Test
    fun trimsTrailingDotsAndSpaces() {
        assertEquals("name", Paths.safeFilename("name.   "))
        assertEquals("name", Paths.safeFilename("name..."))
    }

    @Test
    fun capsTheLength() {
        assertEquals(120, Paths.safeFilename("a".repeat(500) + ".png").length)
    }

    @Test
    fun theResultCanNeverEscapeItsDirectory() {
        val nasty = listOf(
            "../../../etc/passwd",
            """..\..\windows\evil.exe""",
            "....//....//x",
            "/absolute/path",
            "..",
        )
        for (n in nasty) {
            val safe = Paths.safeFilename(n)
            assertTrue(n, !safe.contains('/'))
            assertTrue(n, !safe.contains('\\'))
            assertNotEquals(n, "..", safe)
        }
    }

    @Test
    fun uniquePathUsesThePlainNameWhenFree() {
        val dir = tmp.newFolder()
        assertEquals(java.io.File(dir, "a.png"), Paths.uniquePath(dir, "a.png"))
    }

    @Test
    fun uniquePathNeverOverwritesAnExistingCatch() {
        val dir = tmp.newFolder()
        java.io.File(dir, "photo.png").writeText("first")
        val second = Paths.uniquePath(dir, "photo.png")
        assertEquals(java.io.File(dir, "photo (1).png"), second)

        second.writeText("second")
        assertEquals(java.io.File(dir, "photo (2).png"), Paths.uniquePath(dir, "photo.png"))
        // The original survived, which is the whole point.
        assertEquals("first", java.io.File(dir, "photo.png").readText())
    }

    @Test
    fun uniquePathHandlesNamesWithNoExtension() {
        val dir = tmp.newFolder()
        java.io.File(dir, "README").writeText("x")
        assertEquals(java.io.File(dir, "README (1)"), Paths.uniquePath(dir, "README"))
    }

    @Test
    fun uniquePathTreatsALeadingDotAsPartOfTheName() {
        val dir = tmp.newFolder()
        java.io.File(dir, ".gitignore").writeText("x")
        assertEquals(java.io.File(dir, ".gitignore (1)"), Paths.uniquePath(dir, ".gitignore"))
    }
}
